src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs

```diff
@@ -333,17 +329,14 @@ internal static void RunFromThreadPoolDispatchLoop(ExecutionContext executionCon
 
             // Enregister currentThread to alternative varibale as crossing EH will have forced it to stack
             Thread currentThread1 = currentThread;
-            // The common case is that these have not changed, so avoid the cost of a write barrier if not needed.
-            if (currentThread1.SynchronizationContext != null)
-            {
-                // Restore changed SynchronizationContext back to previous
-                currentThread1.SynchronizationContext = null;
-            }
+
+            // Restore changed SynchronizationContext back to Default
+            currentThread1.SynchronizationContext = null;
```

Comment:
Isn't it better to check for null before writing to it like you had before?

Reply:
Since its `null` constant the write barrier doesn't get involved; so its whether `read+cmp+je` is better than a single non-dependent write (e.g.)
```asm
       49837F0800           cmp      gword ptr [r15+8], 0  ; is null check
       7406                 je       SHORT G_M60392_IG10 ; jump taken
       33C9                 xor      rcx, rcx
       49894F08             mov      gword ptr [r15+8], rcx  ; set null
G_M60392_IG10:
       49837F1000           cmp      gword ptr [r15+16], 0  ; is null check
       7406                 je       SHORT G_M60392_IG11 ; jump taken
       33C9                 xor      rcx, rcx
       49894F10             mov      gword ptr [r15+16], rcx  ; set null
``` 
vs
```asm
       33C9                 xor      rcx, rcx
       49894F08             mov      gword ptr [r15+8], rcx  ; set null
       49894F10             mov      gword ptr [r15+16], rcx  ; set null
```

Reply:
Ah ok, makes sense

Reply:
I'd say the second is quite likely better. 

Stores are generally cheap and branches and loads can be expensive.